### DISCUSSION TITLE
Segmentation fault (core dumped) during tracking

### DISCUSSION THREAD

**Discussion by P-Myers** [WAM2layers/WAM2layers#408] [Q&A] [answered]
Hi everyone,

I am running into a "low level" issue when trying to run the track function of the model. 

First, for some background, I am running the model with the latest **ERA5 data** downloaded from the new Copernicus Beta portal (as the old portal is closing soon). I had to create a script that downloads the necessary data per month, then splits the data into daily files as required by the model. I am more than willing to share this upon request for other users (it has been inspired by user  @pohzim)

Further, the Beta data has a few variable name differences compared to the previous data versions, namely "time" is now "valid_time" and "level" is now "model_level". These must be addressed in order for the preprocessing script to function. I have developed my own adjustment script to rename these variables. I am also willing to share this, however the script is a bit clunky. 

Additionally, I am now discovering that the Beta data utilizes a different time convention than the previous version. This being the Unix Epoch: Seconds since 1970-01-01 00:00:00 UTC, rather than the previous Proleptic Gregorian time (or time since 1900): Hours or days since 1900-01-01 00:00:00 UTC. Converting the preprocess output as well as my raw + adjusted data "time" into human readable form produced the expected dates, leading me to believe that this difference was handled smoothly by the preprocess.

With my splitting, adjustment, and subsequent wam preprocess, I was able to successfully run a very small test track (1 day, 1 hour timestep). I have not been able to recreate this, and foolishly did not save a backup of the conditions that produced this. 

Issue: now, I am continually encountering this "Segmentation fault (core dumped)" which I understand to be at a low level of my system. I have attempted to reproduce the conditions of my successful run, but cannot. The segmentation fault began, I believe, when I attempted to run a larger amount of data (about 1 year) with the model. The preprocess had been successful, but the tracking is where the issue began. 

I am running the model on my universities high powered computer with ample resources, and have all of my packages installed via Docker. I have redownloaded the model and have removed any files from previous runs, but still encounter the issue. I have tried various time configurations for my tagging/tracking, but with still no success. I have attached my configuration file for review. 

After starting fresh, it appeared as if my model attempts were not properly reading my config file, still preprocessing from previous date settings, and even ignoring mistakes intentionally added to the config. This led me to believe that perhaps there is some configuration data cached that is causing my segmentation fault. Could this be?

Has anyone had experience managing similar issues? I have tried many different options, but still have no solution.

I have very little previous experience with python and computer/data science in general, so I apologize if anything is not clear. If there are any further files or information I should provide, please let me know. I am grateful of any help.

Thank you,

Parker

![Screen Shot 2024-09-27 at 3 50 36 AM](https://github.com/user-attachments/assets/c552c440-1355-4cb0-b035-bf049854d921)

**Comment by Peter9192** [comment]
Hi @P-Myers,

Welcome, and thanks for reaching out! This is a very puzzling issue. It sounds like the configuration file you've been editing is not the same one that's used in the call to wam2layers. When you call wam2layers track, it first logs the configuration file that it uses, like so:

```
❯ wam2layers track config-eiffel.yaml
Welcome to WAM2layers.
Starting backward tracking experiment.
Initializing experiment with config file config-eiffel.yaml
```

Can you confirm this file is the one you've been editing? We could add a similar log message to the preprocessing, by adding a log message [here](https://github.com/WAM2layers/WAM2layers/blob/e6eccd4bdd0bd7da0b9166d39f9c699e8b98be21/src/wam2layers/cli.py#L165-L166) like so:

```
logger.info(f"Using configuration file {config_file}")
```
However, I suspect you may have installed wam2layer from PyPI, not from source, is that correct? That's fine, but that would make it a bit harder to modify the code yourself.

Alternatively, can you try to run one of the example cases? See the [quickstart](https://wam2layers.readthedocs.io/en/latest/quickstart.html). If that works, I would expect there is an issue with the new ERA5 data.

## On the new ERA5 data
Thanks for the detailed report of using the era5 data from the new CDS. This is really helpful. I think eventually we want to support this data source out of the box, rather than asking users to go through the manual conversion process like you did. I image we could have 

```
wam2layers preprocess era5  # This would be the old CDS (for backward compatibility)
wam2layers preprocess era5-newcds  # This would be using data from the new CDS
```

I'll open a new issue about that. PS: #409

**Reply by YiyingWang-00** [reply]
I also have this problem in preprocessing, how can I solve this problem? Thank you!

**Comment by ruudvdent** [comment]
No solution I can provide directly, but thanks for reporting the ERA5 time issue indeed!

**Comment by P-Myers** [comment]
Hi everyone,

I appreciate the quick responses here. I have found that the issue is somehow related to my timestep and the tagging / tracking end times, despite my initial setups matching previous examples I have found throughout the documentation here.

I have found that when setting my tagging/tracking end times to "00:00," together with any timestep at or below "3600," the model would produce a segmentation fault (core dumped) output. I assume that this may be due to the model attempting to access out-of-bounds data, resulting in a segmentation fault? This error led me to experiment with using timesteps above 3600 (breaking CFL criterion), with the lowest threshold being 7200. This results in about 90 - 99% of the grid being corrected with stability corrections.

What (sort of) works is using the precise time "23:59," which allows me to set my timestep to the CFL appropriate 3600 (and below). This produces output, however the output is just a single file, something like "backtrack_2017-01-01T00-59.nc" when the tagging/tracking were set to a range like start_date: "2017-01-01T00:00", end_date: "2017-01-03T23:59". Not only does this output not match the configuration, but it also has a high correction percentage (33 - 45 %). I sampled various timesteps, all with similar results.

Could this be related to some bug caused by the new data? Is there any known place where I can still access pre-beta data to test this, and possibly use for my entire experiment? All signs seemingly point to the Beta data set. 

I have successfully completed the sample Eiffel model run on my system. I should note that I am using the full global ERA5 and my region.nc file is a binary mask of my study area (matching the spatial resolution of my data) which I have throughly inspected for flaws.



Thank you for all your work.

Best,
Parker

**Reply by BSchilperoort** [reply]
Hi Parker, as the wam2layers code is just python code, I doubt that anything here would cause a segfault. Trying to reach out-of-bounds data should raise some kind of error. I would suspect something might be wrong with the beta-cds data, which causes the netCDF4 library (which is C or Fortran) to encounter a segmentation fault.

I am not sure how proficient you are at git and python, but you can try the branch associated with this PR: https://github.com/WAM2layers/WAM2layers/pull/401/
That allows you to use Google's [ARCO ERA5 dataset](https://github.com/google-research/arco-era5), which will be downloaded on-the-fly during preprocessing.

**Reply by Peter9192** [reply]
Might have time for a more detailed look later, but first a quick check: when you say "global data", do you mean -90 to 90 deg lat? Usually we recommend not to use the model above ~75deg latitude.

**Comment by P-Myers** [comment]
Thank you all for the support!

Thankfully, the segmentation fault issue seemed to be related to my use of lat 90, -90 rather than the recommended 75, -75.... beginners mistake, however I have learned a lot in the process. 

On another note, I want to point out that the beta CDS for ERA5 data is still very sluggish with download speeds, specifically for model levels. The ARCO ERA5 seems to be a good solution for this issue.

P.S., willing to share my adjustment code for the new variable names produced by the CDS Beta.

Thanks!

**Reply by ruudvdent** [reply, marked as answer]
New CDS issue is now tackled in #429 and the ARCO ERA5 in #401 and #428. Hope this solves the problems. These changes will be incorporated when v4.2 is released. Thanks for sharing your experiences!
